api/src/main/java/javax/enterprise/inject/spi/builder/InjectionPointBuilder.java

```diff
@@ -0,0 +1,243 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package javax.enterprise.inject.spi.builder;
+
+import javax.enterprise.inject.Default;
+import javax.enterprise.inject.spi.AnnotatedField;
+import javax.enterprise.inject.spi.AnnotatedParameter;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.ProcessInjectionPoint;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.Set;
+
+/**
+ * This API is an helper to build a new {@link InjectionPoint} instance.
+ * CDI container must provides an implementation of this interface accessible by calling
+ * {@link ProcessInjectionPoint#injectionPointBuilder()}
+ * An InjectionPointBuilder shouldn't be used after the end of ProcessInjectionPoint observer invocation.
+ *
+ * @author Antoine Sabot-Durand
+ * @since 2.0
+ */
+public interface InjectionPointBuilder {
+
+
+    /**
+     * Read the InjectionPoint information from the given {@link Field}.
+     * All relevant information is overwritten.
+     *
+     * @param field defining the InjectionPoint.
+     * @return self
+     */
+    InjectionPointBuilder read(Field field);
+
+    /**
+     * Read the InjectionPoint information from the given {@link Method}.
+     * All relevant information is overwritten.
+     *
+     * @param method defining the InjectionPoint.
+     * @param param  the parameter's index in the Method defining the InjectionPoint
+     * @return self
+     */
+    InjectionPointBuilder read(Method method, int param);
+
+    /**
+     * Read the InjectionPoint information from the given {@link Constructor}.
+     * All relevant information is overwritten.
+     *
+     * @param constructor defining the InjectionPoint.
+     * @param param       the parameter's index in the Constructor defining the InjectionPoint
+     * @return self
+     */
+    InjectionPointBuilder read(Constructor<?> constructor, int param);
```

Comment:
It might be worth to have another `read` method accepting the parameter name for the constructor argument